rfcs/text/0428-extending-and-duplicating-types.md

```diff
@@ -0,0 +1,970 @@
+
+### Problems with `unevaluatedProperties`
+
+`unevaluatedProperties` _almost_ provides the functionality we're after, but unfortunately it just barely misses. If supertypes themselves specify `{ "unevaluatedProperties": false }`, they are not able to be part of an `allOf` validator, as they will error out as soon as they see properties that are not part of the supertype itself.
+
+When composing schemas that all contain `{ "unevaluatedProperties": false }`, each schema will disallow any other properties which they do not define. Using the following JSON Schema as an example:
+
+```json
+{
+  "allOf": [
+    {
+      "type": "object",
+      "properties": {
+        "city": { "type": "string" }
+      },
+      "required": ["city"],
+      "unevaluatedProperties": false
+    }
+  ],
+  "type": "object",
+  "properties": {
+    "type": { "name": "string" }
+  },
+  "unevaluatedProperties": false
+}
+```
+
+and trying to validate against
+
+```json
+{
+  "city": "Copenhagen",
+  "name": "Charles"
+}
+```
+
+results in validations errors such as `Property 'name' has not been successfully evaluated and the schema does not allow unevaluated properties`.
+
+Changing from `unevaluatedProperties` to `additionalProperties` results in errors `Property 'name' has not been defined and the schema does not allow additional properties`.
+
+Both of these solutions for strict schemas would not be suitable for the type of expressiveness we want for type extension, unfortunately.
+
+The required behavior is that `unevaluatedProperties` should only validate at the top level of a type, allowing supertypes to validate `unevaluatedProperties` but defer checking if they're used within in subtype.
+
+### Redefining `unevaluatedProperties`
+
+We're already implicitly defining `{ "additionalProperties": false }` in the [versioning RFC](./0408-versioning-types.md#determining-type-compatibility) for all schemas, this RFC will piggyback on that, and change the existing, implicit `additionalProperties` usage to `unevaluatedProperties` and slightly change its [vocabulary definition](https://json-schema.org/draft/2020-12/json-schema-core.html#name-unevaluatedproperties).
+
+The vocabulary change is that the `unevaluatedProperties` keyword will only be applicable at top-level schemas (i.e. subtypes or schemas that don't specify `allOf`), and ignored when present on a schema that resides in the `allOf` array. Instead of the default value being `{ "unevaluatedProperties": {} }`, the default value will be `{ "unevaluatedProperties": false }`. Together this makes it so that all supertypes will be able to compose, as their individual, implicit `unevaluatedProperties` values won't have any effect on validation and defer evaluation to the extending type (the subtype).
```

Comment:
I think it would be better to not change the vocabulary of JSON schema and instead just say we _implicitly introduce_ `unevaluatedProperties` to the top level of an entity type when we consider it for validation.

(I think we should do this _if_ we don't split the schemas into constituent parts as suggested for consideration later on)

Reply:
This isn't that different to what's written here, but as long as we just don't include `unevaluatedProperties` within the meta-schemas, it won't be possible (valid) to introduce the keyword in other places, and we don't need to change the vocabulary to say we "ignore it" or otherwise.

Reply:
This has been written into the document and tested locally